aesd-char-driver/main.c

```diff
@@ -6,65 +6,237 @@
+char devices. */
 int aesd_open(struct inode *inode, struct file *filp)
 {
+    struct aesd_dev *dev; /* device information */
     PDEBUG("open");
-    /**
-     * TODO: handle open
-     */
-    return 0;
+    
+    dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+    filp->private_data = dev; /*for other methods*/
+    return 0; /* success */
 }
 
+
 int aesd_release(struct inode *inode, struct file *filp)
 {
     PDEBUG("release");
     /**
-     * TODO: handle release
+     * TODO: handle release. The basic form of aesdchar has no hardware to shutdown
+     * aesd_release() doesn't need to deallocate anything as open() didn't allocate
+     * anything for filp->private_data
      */
     return 0;
 }
 
 ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
                 loff_t *f_pos)
 {
+    /* filp- file pointer: private_data member can be used to get aesd_dev
+        buff - buffer to fill. Can we access this buffer directly? No, use copy_to_user, read 
+        from device basically. count - max number of bytes to write to buff. You 
+        may want/need to write less than this*/
+    //*buf is the buffer to fill, copy from kernel space to user space
+    //count - max number of bytes to write ("read into") buff, 
+    // May want/need to write less than this
+    // also reading the f_pos which is the char_offset??
     ssize_t retval = 0;
+    ssize_t read_bytes;
+    size_t entry_offset_rd;
+    struct aesd_dev *dev = NULL;
+    struct aesd_buffer_entry *entry = NULL;
     PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
-    /**
-     * TODO: handle read
-     */
+
+    if (filp == NULL || buf == NULL)
+    {
+        retval = -EINVAL;
+        PDEBUG("file pointer filp NULL, returning %zu", retval);
+        goto errout;
+    }
+
+    dev = filp->private_data; //device pointer acquired here
+
+    if (mutex_lock_interruptible(&dev->bufferlock))
+    {
+        retval = -ERESTARTSYS;
+        PDEBUG("mutex lock interruptible, returning %zu", retval);
+        goto errout;
+    }
+    
+    entry = aesd_circular_buffer_find_entry_offset_for_fpos(&(dev->circular), *f_pos, &entry_offset_rd);
+
+    if (entry == NULL)
+    {
+        retval = 0;
+        PDEBUG("aesd_circular_buffer_find_entry_offset_for_fpos NULL return %zu", retval);
+        goto out; //mutex is held so can't goto errout
+    }
+
+    //check/get the minimum between count and the size minus the entry offset
+    read_bytes = ((entry->size - entry_offset_rd) < count) ? (entry->size - entry_offset_rd) : count;
+
+    if(copy_to_user(buf, (entry->buffptr + entry_offset_rd), read_bytes))
+    {
+        retval = -EFAULT;
+        PDEBUG("copy_to_user non-negative value, returning %zu", retval);
+        goto out; //mutex is held so can't goto errout
+    }
+
+    //point to the next offset
+    *f_pos += read_bytes;
+    retval = read_bytes;
+    goto out;
+
+    errout:
```

Comment:
@sota6640 should be possible to reorder such that you don't need to do a goto in the success case.